packages/xen-api/src/index.js

```diff
@@ -698,25 +698,32 @@ export class Xapi extends EventEmitter {
 
   _sessionCallRetryOptions = {
     tries: 2,
-    when: error => this._status !== DISCONNECTED && error?.code === 'SESSION_INVALID',
+    when: error =>
+      this._status !== DISCONNECTED && error?.code === 'SESSION_INVALID' && this._auth.password !== undefined,
     onRetry: () => this._sessionOpen(),
   }
-  _sessionCall(method, args, timeout) {
+  async _sessionCall(method, args, timeout) {
     if (method.startsWith('session.')) {
       return Promise.reject(new Error('session.*() methods are disabled from this interface'))
     }
 
-    return pRetry(() => {
-      const sessionId = this._sessionId
-      assert.notStrictEqual(sessionId, undefined)
+    try {
+      return await pRetry(() => {
+        const sessionId = this._sessionId
+        assert.notStrictEqual(sessionId, undefined)
 
-      const newArgs = [sessionId]
-      if (args !== undefined) {
-        newArgs.push.apply(newArgs, args)
-      }
+        const newArgs = [sessionId]
+        if (args !== undefined) {
+          newArgs.push.apply(newArgs, args)
+        }
+
+        return this._call(method, newArgs, timeout)
+      }, this._sessionCallRetryOptions)
+    } catch (error) {
+      await ignoreErrors.call(this.disconnect())
```

Comment:
Why should we disconnect each time a call fails??

Reply:
I forgot to add the `error.code` test again after changing my mind a few times on the implementation.